### PR TITLE
[Merged by Bors] - feat: port `Algebra.GCDMonoid.Basic`

### PR DESCRIPTION
**Pull request by winstonyin**

### CODE REVIEW

**Comment by winstonyin** [comment]
I've marked those proofs which are taking a while to compile, if somebody wants to optimise them for speed.

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,1438 @@
...
+-- @[protect_proj]
+class NormalizationMonoid (α : Type _) [CancelCommMonoidWithZero α] where
+  /-- `normUnit` assigns to each element of the monoid a unit of the monoid. -/
+  normUnit : α → αˣ
+  /-- The proposition that `normUnit` maps `0` to the identity. -/
+  normUnit_zero : normUnit 0 = 1
+  /-- The proposition that `normUnit` respects multiplication of non-zero elements. -/
+  normUnit_mul : ∀ {a b}, a ≠ 0 → b ≠ 0 → normUnit (a * b) = normUnit a * normUnit b
+  /-- The proposition that `normUnit` maps units to their inverses. -/
+  normUnit_coe_units : ∀ u : αˣ, normUnit u = u⁻¹
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:85]:
```suggestion
class NormalizationMonoid (α : Type _) [CancelCommMonoidWithZero α] where
  /-- `normUnit` assigns to each element of the monoid a unit of the monoid. -/
  protected normUnit : α → αˣ
  /-- The proposition that `normUnit` maps `0` to the identity. -/
  protected normUnit_zero : normUnit 0 = 1
  /-- The proposition that `normUnit` respects multiplication of non-zero elements. -/
  protected normUnit_mul : ∀ {a b}, a ≠ 0 → b ≠ 0 → normUnit (a * b) = normUnit a * normUnit b
  /-- The proposition that `normUnit` maps units to their inverses. -/
  protected normUnit_coe_units : ∀ u : αˣ, normUnit u = u⁻¹
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1438 @@
...
+  /-- The GCD is a divisor of the second element. -/
+  gcd_dvd_right : ∀ a b, gcd a b ∣ b
+  /-- Tny common divisor of both elements is a divisor of the GCD. -/
+  dvd_gcd : ∀ {a b c}, a ∣ c → a ∣ b → a ∣ gcd c b
+  /-- The product of two elements is `Associated` with the product of their GCD and LCM. -/
+  gcd_mul_lcm : ∀ a b, Associated (gcd a b * lcm a b) (a * b)
+  /-- `0` is left-absorbing. -/
+  lcm_zero_left : ∀ a, lcm 0 a = 0
+  /-- `0` is right-absorbing. -/
+  lcm_zero_right : ∀ a, lcm a 0 = 0
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:291]:
```suggestion
class GCDMonoid (α : Type _) [CancelCommMonoidWithZero α] where
  /-- The greatest common divisor between two elements. -/
  protected gcd : α → α → α
  /-- The least common multiple between two elements. -/
  protected lcm : α → α → α
  /-- The GCD is a divisor of the first element. -/
  protected gcd_dvd_left : ∀ a b, gcd a b ∣ a
  /-- The GCD is a divisor of the second element. -/
  protected gcd_dvd_right : ∀ a b, gcd a b ∣ b
  /-- Tny common divisor of both elements is a divisor of the GCD. -/
  protected dvd_gcd : ∀ {a b c}, a ∣ c → a ∣ b → a ∣ gcd c b
  /-- The product of two elements is `Associated` with the product of their GCD and LCM. -/
  protected gcd_mul_lcm : ∀ a b, Associated (gcd a b * lcm a b) (a * b)
  /-- `0` is left-absorbing. -/
  protected lcm_zero_left : ∀ a, lcm 0 a = 0
  /-- `0` is right-absorbing. -/
  protected lcm_zero_right : ∀ a, lcm a 0 = 0
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,1438 @@
...
+
+export GCDMonoid (gcd lcm gcd_dvd_left gcd_dvd_right dvd_gcd lcm_zero_left lcm_zero_right)
+
+attribute [simp] lcm_zero_left lcm_zero_right
+
+section GCDMonoid
+
+variable [CancelCommMonoidWithZero α]
+
+-- Porting note: linter wanted the lhs in this simp-normal form instead of `normalize (gcd a b)`
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:316]: Hmm, I'm not sure that's the best solution

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,1438 @@
...
+#align dvd_mul_gcd_of_dvd_mul dvd_mul_gcd_of_dvd_mul
+
+/-- Represent a divisor of `m * n` as a product of a divisor of `m` and a divisor of `n`.
+
+In other words, the nonzero elements of a `GCDMonoid` form a decomposition monoid
+(more widely known as a pre-Schreier domain in the context of rings).
+
+Note: In general, this representation is highly non-unique.
+
+See `nat.prod_dvd_and_dvd_of_dvd_prod` for a constructive version on `ℕ`.  -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:527]:
```suggestion
See `Nat.prodDvdAndDvdOfDvdProd` for a constructive version on `ℕ`.  -/
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,1438 @@
...
+    exact
+      (gcd_zero_left' (0 ^ k : α)).dvd.trans
+        (pow_dvd_pow_of_dvd (gcd_zero_left' (0 : α)).symm.dvd _)
+  · induction' k with k hk
+    · rw [pow_zero, pow_zero]
+      exact (gcd_one_right' a).dvd
+    rw [pow_succ, pow_succ]
+    trans gcd a b * gcd a (b ^ k)
+    apply gcd_mul_dvd_mul_gcd a b (b ^ k)
+    exact (mul_dvd_mul_iff_left hg).mpr hk
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:578]:
```suggestion
    · apply gcd_mul_dvd_mul_gcd a b (b ^ k)
    · exact (mul_dvd_mul_iff_left hg).mpr hk
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,1438 @@
...
+#align extract_gcd extract_gcd
+
+end GCD
+
+section LCM
+
+theorem lcm_dvd_iff [GCDMonoid α] {a b c : α} : lcm a b ∣ c ↔ a ∣ c ∧ b ∣ c := by
+  by_cases a = 0 ∨ b = 0
+  ·
+    rcases h with (rfl | rfl) <;>
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:707]:
```suggestion
  · rcases h with (rfl | rfl) <;>
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,1438 @@
...
+private theorem map_mk_unit_aux [DecidableEq α] {f : Associates α →* α}
+    (hinv : Function.RightInverse f Associates.mk) (a : α) :
+    a * ↑(Classical.choose (associated_map_mk hinv a)) = f (Associates.mk a) :=
+  Classical.choose_spec (associated_map_mk hinv a)
+
+/-- Define `NormalizationMonoid` on a structure from a `MonoidHom` inverse to `Associates.mk`. -/
+def normalizationMonoidOfMonoidHomRightInverse [DecidableEq α] (f : Associates α →* α)
+    (hinv : Function.RightInverse f Associates.mk) :
+    NormalizationMonoid
+      α where
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1025]:
```suggestion
    NormalizationMonoid α where
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 8]
```diff
@@ -0,0 +1,1438 @@
...
+variable (G₀ : Type _) [CommGroupWithZero G₀] [DecidableEq G₀]
+
+-- Porting note: very slow; improve performance?
+-- see Note [lower instance priority]
+instance (priority := 100) :
+    NormalizedGCDMonoid
+      G₀ where
+  normUnit x := if h : x = 0 then 1 else (Units.mk0 x h)⁻¹
+  normUnit_zero := dif_pos rfl
+  normUnit_mul := @fun x y x0 y0 => Units.eq_iff.1 (by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1375]:
```suggestion
  normUnit_mul {x y} x0 y0 := Units.eq_iff.1 (by
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 9]
```diff
@@ -0,0 +1,1438 @@
...
+#align normalization_monoid_of_monoid_hom_right_inverse normalizationMonoidOfMonoidHomRightInverse
+
+/-- Define `GCDMonoid` on a structure just from the `gcd` and its properties. -/
+noncomputable def gcdMonoidOfGCD [DecidableEq α] (gcd : α → α → α)
+    (gcd_dvd_left : ∀ a b, gcd a b ∣ a) (gcd_dvd_right : ∀ a b, gcd a b ∣ b)
+    (dvd_gcd : ∀ {a b c}, a ∣ c → a ∣ b → a ∣ gcd c b) : GCDMonoid α :=
+  { gcd
+    gcd_dvd_left
+    gcd_dvd_right
+    dvd_gcd := @fun a b c => dvd_gcd
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1060]:
```suggestion
    dvd_gcd := fun {a b c} => dvd_gcd
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 10]
```diff
@@ -0,0 +1,1438 @@
...
+/-- Define `NormalizedGCDMonoid` on a structure just from the `gcd` and its properties. -/
+noncomputable def normalizedGCDMonoidOfGCD [NormalizationMonoid α] [DecidableEq α] (gcd : α → α → α)
+    (gcd_dvd_left : ∀ a b, gcd a b ∣ a) (gcd_dvd_right : ∀ a b, gcd a b ∣ b)
+    (dvd_gcd : ∀ {a b c}, a ∣ c → a ∣ b → a ∣ gcd c b)
+    (normalize_gcd : ∀ a b, normalize (gcd a b) = gcd a b) : NormalizedGCDMonoid α :=
+  { (inferInstance : NormalizationMonoid α) with
+    gcd
+    gcd_dvd_left
+    gcd_dvd_right
+    dvd_gcd := @fun a b c => dvd_gcd
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1093]:
```suggestion
    dvd_gcd := fun {a b c} => dvd_gcd
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 11]
```diff
@@ -0,0 +1,1438 @@
...
+        intro con
+        have h := lcm_dvd (Dvd.intro b rfl) (Dvd.intro_left a rfl)
+        rw [con, zero_dvd_iff, mul_eq_zero] at h
+        cases h
+        · exact absurd ‹a = 0› h
+        · exact absurd ‹b = 0› h_1
+      rw [← mul_dvd_mul_iff_left h0, ← Classical.choose_spec (exists_gcd a b),
+        mul_dvd_mul_iff_right h_1]
+      apply dvd_lcm_left
+    dvd_gcd := @fun a b c ac ab => by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1200]:
```suggestion
    dvd_gcd := fun {a b c} ac ab => by
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 12]
```diff
@@ -0,0 +1,1438 @@
...
+        intro con
+        have h := lcm_dvd (Dvd.intro b rfl) (Dvd.intro_left a rfl)
+        rw [con, zero_dvd_iff, mul_eq_zero] at h
+        cases h
+        · exact absurd ‹a = 0› h
+        · exact absurd ‹b = 0› h_1
+      rw [← mul_dvd_mul_iff_left h0, ← Classical.choose_spec (exists_gcd a b), normalize_dvd_iff,
+        mul_dvd_mul_iff_right h_1]
+      apply dvd_lcm_left
+    dvd_gcd := @fun a b c ac ab => by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1295]:
```suggestion
    dvd_gcd := fun {a b c} ac ab => by
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 13]
```diff
@@ -0,0 +1,1438 @@
...
+      apply ac }
+#align normalized_gcd_monoid_of_lcm normalizedGCDMonoidOfLCM
+
+/-- Define a `GCDMonoid` structure on a monoid just from the existence of a `gcd`. -/
+noncomputable def gcdMonoidOfExistsGCD [DecidableEq α]
+    (h : ∀ a b : α, ∃ c : α, ∀ d : α, d ∣ a ∧ d ∣ b ↔ d ∣ c) : GCDMonoid α :=
+  gcdMonoidOfGCD (fun a b => Classical.choose (h a b))
+    (fun a b => ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).1)
+    (fun a b => ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).2)
+    @fun a b c ac ab => (Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1326]:
```suggestion
    fun {a b c} ac ab => (Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 14]
```diff
@@ -0,0 +1,1438 @@
...
+
+/-- Define a `NormalizedGCDMonoid` structure on a monoid just from the existence of a `gcd`. -/
+noncomputable def normalizedGCDMonoidOfExistsGCD [NormalizationMonoid α] [DecidableEq α]
+    (h : ∀ a b : α, ∃ c : α, ∀ d : α, d ∣ a ∧ d ∣ b ↔ d ∣ c) : NormalizedGCDMonoid α :=
+  normalizedGCDMonoidOfGCD (fun a b => normalize (Classical.choose (h a b)))
+    (fun a b =>
+      normalize_dvd_iff.2 ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).1)
+    (fun a b =>
+      normalize_dvd_iff.2 ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).2)
+    (@fun a b c ac ab => dvd_normalize_iff.2 ((Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩))
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1337]:
```suggestion
    (fun {a b c} ac ab => dvd_normalize_iff.2 ((Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩))
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 15]
```diff
@@ -0,0 +1,1438 @@
...
+    fun _ _ => normalize_idem _
+#align normalized_gcd_monoid_of_exists_gcd normalizedGCDMonoidOfExistsGCD
+
+/-- Define a `GCDMonoid` structure on a monoid just from the existence of an `lcm`. -/
+noncomputable def gcdMonoidOfExistsLCM [DecidableEq α]
+    (h : ∀ a b : α, ∃ c : α, ∀ d : α, a ∣ d ∧ b ∣ d ↔ c ∣ d) : GCDMonoid α :=
+  gcdMonoidOfLCM (fun a b => Classical.choose (h a b))
+    (fun a b => ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).1)
+    (fun a b => ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).2)
+    @fun a b c ac ab => (Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1347]:
```suggestion
    fun {a b c} ac ab => (Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 16]
```diff
@@ -0,0 +1,1438 @@
...
+
+/-- Define a `NormalizedGCDMonoid` structure on a monoid just from the existence of an `lcm`. -/
+noncomputable def normalizedGCDMonoidOfExistsLCM [NormalizationMonoid α] [DecidableEq α]
+    (h : ∀ a b : α, ∃ c : α, ∀ d : α, a ∣ d ∧ b ∣ d ↔ c ∣ d) : NormalizedGCDMonoid α :=
+  normalizedGCDMonoidOfLCM (fun a b => normalize (Classical.choose (h a b)))
+    (fun a b =>
+      dvd_normalize_iff.2 ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).1)
+    (fun a b =>
+      dvd_normalize_iff.2 ((Classical.choose_spec (h a b) (Classical.choose (h a b))).2 dvd_rfl).2)
+    (@fun a b c ac ab => normalize_dvd_iff.2 ((Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩))
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1358]:
```suggestion
    (fun {a b c} ac ab => normalize_dvd_iff.2 ((Classical.choose_spec (h c b) a).1 ⟨ac, ab⟩))
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 17]
```diff
@@ -0,0 +1,1438 @@
...
+    split_ifs with h
+    · rw [h.1]
+    · exact one_dvd _
+  gcd_dvd_right a b := by
+    -- Porting note: need `dsimp only`
+    dsimp only
+    split_ifs with h
+    · rw [h.2]
+    · exact one_dvd _
+  dvd_gcd := @fun a b c hac hab => by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1403]:
```suggestion
  dvd_gcd {a b c} hac hab := by
```

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 18]
```diff
@@ -0,0 +1,1438 @@
...
+        exact (associated_one_iff_isUnit.mpr ((IsUnit.mk0 _ ha).mul (IsUnit.mk0 _ hb))).symm
+  lcm_zero_left b := if_pos (Or.inl rfl)
+  lcm_zero_right a := if_pos (Or.inr rfl)
+  -- `split_ifs` wants to split `normalize`, so handle the cases manually
+  normalize_gcd a b := if h : a = 0 ∧ b = 0 then by simp [if_pos h] else by simp [if_neg h]
+  normalize_lcm a b := if h : a = 0 ∨ b = 0 then by simp [if_pos h] else by simp [if_neg h]
+
+@[simp]
+theorem coe_norm_unit {a : G₀} (h0 : a ≠ 0) : (↑(normUnit a) : G₀) = a⁻¹ := by simp [normUnit, h0]
+#align comm_group_with_zero.coe_norm_unit CommGroupWithZero.coe_norm_unit
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:1433]:
```suggestion
theorem coe_normUnit {a : G₀} (h0 : a ≠ 0) : (↑(normUnit a) : G₀) = a⁻¹ := by simp [normUnit, h0]
#align comm_group_with_zero.coe_norm_unit CommGroupWithZero.coe_normUnit
```

**Review by Ruben-VandeVelde — CHANGES_REQUESTED** [review]

File: Mathlib/Algebra/GCDMonoid/Basic.lean (hunk 3 above)
Comment (winstonyin) [reply Mathlib/Algebra/GCDMonoid/Basic.lean:316]: I made a thread on Zulip about this: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simp.20normal.20form/near/318018031

Comment (winstonyin) [reply Mathlib/Algebra/GCDMonoid/Basic.lean]: I ended up keeping the original lemma statements and using a lower `simp` priority on them. Hope this is resolved.

File: Mathlib/Algebra/GCDMonoid/Basic.lean [hunk 19]
```diff
@@ -0,0 +1,1434 @@
...
+    have := Subsingleton.elim a_norm b_norm
+    subst this
+    rfl⟩
+#align
+  subsingleton_normalized_gcd_monoid_of_unique_units
+  subsingleton_normalizedGCDMonoid_of_unique_units
+
+@[simp]
+theorem norm_unit_eq_one (x : α) : normUnit x = 1 :=
+  rfl
```
Comment (dupuisf) [inline Mathlib/Algebra/GCDMonoid/Basic.lean:959]:
```suggestion
theorem normUnit_eq_one (x : α) : normUnit x = 1 :=
  rfl
```

**Review by dupuisf — COMMENTED** [review]
Thanks!

bors d+

**Pull request closed without merging**